python/extensions/python.bzl

```diff
@@ -17,40 +17,132 @@
-            # We collect all of the toolchain names to create
-            # the INTERPRETER_LABELS map.  This is used
-            # by interpreter_extensions.bzl
-            toolchains.append(toolchain_attr.name)
+            # If we are in the root module we always register the toolchain.
+            # We wait to register the default toolchain till the end.
+            if mod.is_root:
+                toolchain_names.append(toolchain_attr.name)
+
+                # If we have the default version or we only have one toolchain
+                # in the root module we set the toolchain as the default toolchain.
+                if toolchain_attr.default_version or len(mod.tags.toolchain) == 1:
+                    # We have already found one default toolchain, and we can
+                    # only have one.
+                    if default_toolchain != None:
+                        fail("""We found more than one toolchain that is marked 
+as the default version.  Only set one toolchain with default_version set as 
+True.""")
+                    default_toolchain = toolchain_attr
+                    continue
+
+                #  Always register toolchains that are in the root module.
+                _python_register_toolchains(toolchain_attr, True)
+            else:
+                # We add the toolchain to a map, and we later create the
+                # toolchain if the root module does not have a toolchain with
+                # the same name.  We have to loop through all of the modules to
+                # ensure that we get a full list of the root toolchains.
+                sub_toolchains_map[toolchain_attr.name] = toolchain_attr
+
+    # We did not find a default toolchain so we fail.
+    if default_toolchain == None:
+        fail("""Unable to find a default toolchain in the root module.  
+Please define a toolchain that has default_version set to True.""")
+
+    # Create the toolchains in the submodule(s).
+    for name, toolchain_attr in sub_toolchains_map:
+        # A sub module cannot have a toolchain that is marked as the
+        # default version. TODO: should we create the toolchain anyways,
+        # but set the default version to False?
+        if toolchain_attr.default_version:
+            fail("""Not able to create toolchain named: {}.  This toolchain exists
+in a sub module and defalult_version is set to True.""".format(name))
 
+        # We cannot have a toolchain in a sub module that has the same name of
+        # a toolchain in the root module. This will cause name clashing.
+        if name in toolchain_names:
+            _print_warn("""Not creating the toolchain from sub module, with the name {}. The root
+ modhas a toolchain of the same name.""".format(toolchain_attr.name))
+            continue
+        toolchain_names.append(name)
+        _python_register_toolchains(toolchain_attr, True)
+
+    # We register the default toolchain last.
+    _python_register_toolchains(default_toolchain, False)
```

Comment:
There's two aspects to this: ordering, and compatiblity.

Usually the order targets are defined in doesn't matter, but toolchains are an exception. Basically, all the `native.toolchain` targets that are found via `register_toolchains()` calls are searched in the order they're encountered.

For example, say we had this build file:

```
toolchain(name="foo", ...)
toolchain(name="bar", ...)
```

And this module:

```
register_toolchains("//:all")
```

This will record a list of `[foo, bar]`. When someone asks for a toolchain, it'll first check `foo`, and, if it's compatible, will use it. Otherwise it moves onto the next item in the list.

If we change the BUILD file to put "bar" before "foo", bar will be checked first. If we change the register_toolchains call to explicitly name the targets, e.g. `register_toolchains("//:bar", "//:foo")`, that will also make bar checked first.

Next, compatibility: when iterating over the toolchains, the first compatible toolchain is used, where "compatible" means "are all of the toolchain's constraints satisfied?". For example, say we had this build file:

```
toolchain(name="default_linux", compatible_with=["//os:linux"])
toolchain(name="py38_linux", compatible_with=["//os:linux", "//py:py38"])
```

And we do `register_toolchains("//:all")`, so the ordering is `[default_linux, py38_linux]`. Note how the first toolchain has fewer constraints than the second -- it just cares about matching linux.

When a multi-version aware rule is used, it's constraints are `(os=linux, py=py38)`. We want it to use the `py38_linux` toolchain. But, `default_linux` is checked first and will be matched because all its constraints (it just needs linux) are satisfied, and thus it uses the wrong toolchain.

The toolchain docs explain the algorithm in detail: https://bazel.build/extending/toolchains#toolchain-resolution

Reply:
I think that targets within a package are ordered lexicographically, not in declaration order. That's why rules_go appends a prefix such as `0001_` to all toolchain definitions.

Reply:
You're right! Is that documented somewhere? I couldn't find a mention of `:all` or the lexicographic ordering it uses.

Reply:
I don't think so, @Wyverald dug this out of the code when I asked him about it.

Reply:
https://cs.opensource.google/bazel/bazel/+/refs/heads/master:src/main/java/com/google/devtools/build/lib/packages/Package.java;drc=665a3241f4acf2291d157c76a413d872ae59241e;l=704

All the target pattern resolution stuff eventually go to `Package.getTargets()`, and it's an `ImmutableSortedSet` on the target name.

Reply:
@chrislovecnm No, we don't need to address all of this in this PR because there isn't a hub for rules_python to call register_toolchains on.

The changes for this PR are:
 1. Remove L99, the python_register_toolchains() call for the default toolchain.
 2. Remove the skip-if-default logic up around L56 (so that the repo for the default version is still created)

When we do the PR to automatically register toolchains, then we'll need to worry about what this comment thread is about.